wolfcrypt/src/port/Renesas/renesas_tsip_aes.c

```diff
@@ -151,6 +168,349 @@ int wc_tsip_AesCbcDecrypt(struct Aes* aes, byte* out, const byte* in, word32 sz)
+        XFREE(aadBuf,    aes->heap, DYNAMIC_TYPE_AES);
+    
+        tsip_hw_unlock();
+    }
+    return ret;
+}
+/* 
+ * Decrypt cipher data into plaindata and output authentication tag data.
+ * The session key used for decryption is generated inside this function and 
+ * the key which has been generated and stored in Aes is not used.
+ * parameter
+ *  - aes:  Aes structure
+ *  - out:  buffer where the plain text is output
+ *  - in:   buffer where the cipher data is storead
+ *  - sz:   size of cipher data and also means output size
+ *  - iv:   iv should be consist of implicit-iv of 4 bytes and exp-iv of 8 bytes
+ *  - authTag: buffer where the authentication data is stored.
+ *  - authTagSz: buffer size for authentication data.
+ *  - authIn: buffer where Additionnal Authentication Data(AAD) is stored
+ *  - authInSz: AAD size
+ * return 0 on success, otherwise on error. 
+ * Note: As of TSIPv1.13, only accept 128 and 256 bit of key size 
+ * 
+ */
+int wc_tsip_AesGcmDecrypt(
+    Aes*        aes,      byte* out,
+    const byte* in,       word32 sz,
+    const byte* iv,       word32 ivSz,
+    const byte* authTag,  word32 authTagSz,
+    const byte* authIn,   word32 authInSz,
+          void* ctx)
+{
+    int                 ret = -1;
+    e_tsip_err_t        err;
+    tsip_gcm_handle_t   hdl;
+
+    uint32_t            dataLen = sz;
+    uint32_t            plainBufSz;
+
+    aesGcmDecInitFn     initFn;
+    aesGcmDecUpdateFn   updateFn;
+    aesGcmDecFinalFn    finalFn;
+
+    uint8_t* cipherBuf = NULL;
+    uint8_t* plainBuf  = NULL;
+    uint8_t* aTagBuf   = NULL;
+    uint8_t* aadBuf    = NULL;
+
+    tsip_aes_key_index_t key_server_aes;
+    TsipUserCtx *userCtx;
+
+    WOLFSSL_MSG(">> wc_tsip_AesGcmDecrypt");
+
+    if (aes == NULL || in == NULL || out == NULL || sz == 0 || ctx == NULL ||
+        (ivSz < AESGCM_NONCE_SZ && iv != NULL) ||  /* Requires 12 bytes of iv */
+        (authInSz != 0 && authIn == NULL) ||
+        (authInSz == 0 && authIn != NULL) ||
+        (authTagSz != 0 && authTag == NULL) ||
+        (authTagSz == 0 && authTag != NULL)) 
+        return BAD_FUNC_ARG;
+
+    if (aes->ctx.keySize != 16 && aes->ctx.keySize != 32) {
+        WOLFSSL_MSG("<< wc_tsip_AesGcmDecrypt: illegal key size");
+        return  BAD_FUNC_ARG;
+    }
+
+    if (aes->ctx.keySize == 16) {
+        initFn   = R_TSIP_Aes128GcmDecryptInit;
+        updateFn = R_TSIP_Aes128GcmDecryptUpdate;
+        finalFn  = R_TSIP_Aes128GcmDecryptFinal;
+    }
+    else {
+        initFn   = R_TSIP_Aes256GcmDecryptInit;
+        updateFn = R_TSIP_Aes256GcmDecryptUpdate;
+        finalFn  = R_TSIP_Aes256GcmDecryptFinal;
+    }
+
+    userCtx = (TsipUserCtx *)ctx;
 
+    /* buffer for plain data outout must be multiple of AES_BLOCK_SIZE */
+    plainBufSz = ((sz / AES_BLOCK_SIZE) + 1) * AES_BLOCK_SIZE;
+
+    if ((ret = tsip_hw_lock()) == 0) {
+
+        /* allocate buffers for plain-taxt, cipher-text, authTag and AAD.
```

Comment:
Spelling - "plain-taxt" should be "plaintext"

Reply:
Please address

Reply:
Fixed it.